### PR TITLE
SparklineCell: Display absolute value

### LINKED ISSUE #68541: Chart cell: add numerical value display
Add ability to display numeric value next to the sparklien for chart cells: 
image

- Time series table transform should allow optionally configuring what value to pick for each data frame containing field: last, avg, median
- ~~Stretch: also allow picking value from another numerical field (and removing), thereby allowing separate query to fetch the value?~~
- Allow configuring unit in field config
- Table should be sortable by chart cell column. Likely requires inventing `ValueAndDataFrame` field type
- Values & sparklines in a column should be neatly aligned

### PR DESCRIPTION
**Pull request by domasx2**
**What is this feature?**

- Adds value readout to sparkline cell
- Configurable via transform to be last / average / median
- Sparkline columns now properly sortable by this value
- Value can be hidden via "hide value" toggle on sparkline cell type options

image
image
image

https://github.com/grafana/grafana/assets/847684/751ad551-25f7-4df6-8965-3f7ce1c54d05

**Why do we need this feature?**

Sparklines indicate relative trend, this allows user to get a sense of the absolute value. Also makes the column sortable in a sensible way.

**Who is this feature for?**

Table panel users. 

**Which issue(s) does this PR fix?**:

Fixes #68541

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-data/src/types/dataFrame.ts [hunk 1]
```diff
@@ -244,6 +244,11 @@ export interface DataFrame extends QueryResultBase {
   length: number;
 }
 
+// Data frame that include aggregate value, for use by timeSeriesTableTransformer / chart cell type
+export interface DataFrameWithValue extends DataFrame {
+  value: number | null;
+}
+
```
Comment (leeoniya) [inline packages/grafana-data/src/types/dataFrame.ts:251]: @ryantxu WDYT?

File: public/app/features/transformers/timeSeriesTable/timeSeriesTableTransformer.ts [hunk 2]
```diff
@@ -125,3 +136,35 @@ function getLabelFields(frames: DataFrame[]): Record<string, Record<string, Fiel
...
+      if (sortedValues.length > 0) {
+        if (sortedValues.length % 2 === 0) {
+          return (sortedValues[sortedValues.length / 2 - 1] + sortedValues[sortedValues.length / 2]) / 2;
+        }
+        const middleIndex = Math.floor(sortedValues.length / 2);
+        return sortedValues[middleIndex];
+      }
+      return null;
+  }
+}
```
Comment (torkelo) [inline public/app/features/transformers/timeSeriesTable/timeSeriesTableTransformer.ts:170]: I think you should just be able reuse our existing reducers here, fieldReducers.get(valueType)  , make valueType a  ReducerID

File: public/app/features/transformers/timeSeriesTable/TimeSeriesTableTransformEditor.tsx [hunk 3]
```diff
@@ -1,19 +1,63 @@
...
+      {refIds.map((refId) => {
+        return (
+          <div key={refId}>
+            <InlineFieldRow>
+              <InlineField label={`Trend ${refIds.length > 1 ? ` #${refId}` : ''} value`}>
+                <Select
+                  options={valueTypeOptions}
+                  value={options.refIdToValueType?.[refId] ?? ValueType.Last}
+                  onChange={onSelectValueType.bind(undefined, refId)}
+                />
```
Comment (torkelo) [inline public/app/features/transformers/timeSeriesTable/TimeSeriesTableTransformEditor.tsx:45]: use StatsPicker here (select a reducer),

File: public/app/features/transformers/timeSeriesTable/TimeSeriesTableTransformEditor.tsx [hunk 4]
```diff
@@ -1,19 +1,63 @@
...
+      })}
+    </>
+  );
 }
 
+const valueTypeOptions: Array<SelectableValue<ValueType>> = [
+  { value: ValueType.Last, label: 'Last' },
+  { value: ValueType.Average, label: 'Average' },
+  { value: ValueType.Median, label: 'Median' },
+];
```
Comment (torkelo) [inline public/app/features/transformers/timeSeriesTable/TimeSeriesTableTransformEditor.tsx:59]: no need for this, just use StatsPicker

**Review by torkelo — COMMENTED** [review]
I think this looks great! :rocket: :heart: 

Very close to mergable, think just needs to reuse the existing picker and reducer system we have for calculating the value

File: public/app/features/transformers/timeSeriesTable/timeSeriesTableTransformer.ts (hunk 2 above)
Comment (domasx2) [reply public/app/features/transformers/timeSeriesTable/timeSeriesTableTransformer.ts]: ahh, missed this thingys existence :D done!

**Review by oscarkilhed — COMMENTED** [review]
Nice! Have tried it out and it works. There's a small problem if you select all values or all unique values as the reducer. We can either handle the overflow or somehow filter these two stats from the stats select :shrug: 

image

File: packages/grafana-data/src/types/dataFrame.ts (hunk 1 above)
Comment (torkelo) [reply packages/grafana-data/src/types/dataFrame.ts:251]: I think this makes sense. I think the type should probably allow for string values and booleans, not inconceivable to want a sparkline of of booleans states :)

**Review by torkelo — APPROVED** [review]
might be time to remove the feature toggle for the transform (and add some docs for it)

File: packages/grafana-data/src/types/dataFrame.ts (hunk 1 above)
Comment (domasx2) [reply packages/grafana-data/src/types/dataFrame.ts:251]: Maybe idea for the future :) currently sparkline component only supports numbers

Comment (torkelo) [reply packages/grafana-data/src/types/dataFrame.ts:251]: Ah, let's wait then :)

**Comment by domasx2** [comment]
> Nice! Have tried it out and it works. There's a small problem if you select all values or all unique values as the reducer. We can either handle the overflow or somehow filter these two stats from the stats select 🤷

Good point, filtered out these values

**Comment by domasx2** [comment]
> might be time to remove the feature toggle for the transform (and add some docs for it)

:+1:  removed feature toggle, updated docs

**Comment by skeagy** [comment]
Thank you! Sparklines in table cells along with selectable scalar values (Last/MEan/Max) has been a game-changing feature for me. It unlocks the possibility to create a more robust network monitoring application front-end based on Grafana dashboards.

Toward that end, I've opened a bunch of enhancement requests to further make table cell sparklines more useful:

#81295

#81325 

#81391 

#81527 

#94824

#94826

**Pull request merged**